app/forms/user_piv_cac_setup_form.rb

```diff
@@ -0,0 +1,91 @@
+class UserPivCacSetupForm
+  extend Forwardable
+
+  State = Struct.new(:x509_dn_uuid, :x509_dn, :error, :success)
+
+  def initialize(user, token)
+    @state = State.new
+    @user = user
+    @token = token
+  end
+
+  def_delegators :@state, :x509_dn_uuid, :x509_dn, :error, :success
+
+  def submit
+    @state.success = valid_token?
+
+    process_valid_submission if success
+
+    FormResponse.new(success: success, errors: {})
+  end
+
+  private
+
+  attr_reader :user, :token
+
+  def process_valid_submission
+    user.x509_dn_uuid = x509_dn_uuid
+    user.save!
+    Event.create(user_id: user.id, event_type: :piv_cac_enabled)
+  rescue PG::UniqueViolation
+    @state.success = false
+    @state.error = 'piv_cac.already_associated'
+  end
+
+  def valid_token?
+    user_has_no_piv_cac &&
+      token_provided &&
+      token_decoded &&
+      not_error_token &&
+      piv_cac_not_already_associated
+  end
+
+  def token_decoded
+    @data = PivCacService.decode_token(@token)
+    if @data
+      true
+    else
+      @state.error = 'token.invalid'
+      false
```

Comment:
Code Climate says this is not tested.

Reply:
I had removed it in the next PR, but I'll do that here as well since I'm already making some changes.